libutils/file_lib.c

```diff
@@ -665,6 +672,13 @@ int safe_chown(const char *path, uid_t owner, gid_t group)
     int fd = safe_open(path, 0);
     if (fd < 0)
     {
+        // Do additional checking for FIFOs
+        struct stat st;
+        if (stat(path, &st) == 0 && S_ISFIFO(st.st_mode))
+        {
+            return chown(path, owner, group);
+        }
```

Comment:
> Unfortunately, the problem can't be fixed for chmod without introducing a security hole, since the ``fchmodat`` call doesn't honor ``AT_SYMLINK_NOFOLLOW``.

Actually, it can be fixed if we resort to dropping effective privileges temporarily to the owner of the fifo, and group "nobody". A user can still switch the file with a symlink, but it will only succeed if he points it at one of his own files.

But this is more work. I'm perfectly ok with keeping only the part that fixes the hanging.